[호두] 회원가입 미션 3, 4단계 제출합니다.

Junyoung-WON opened the pull request:
안녕하세요, 꼬상!! 🐿️
3, 4단계 어찌저찌 구현해서 제출하도록 하겠습니다!
4단계의 프로그래밍 요구사항 중 하나인 **_"모든 로직에 테스트 코드를 추가한다."_** 라는 요구사항을 만족시키지 못한 것 같습니다. 😥
유효성을 스스로 검증하는 도메인(`Username`, `Email`, `Password`)에 대한 테스트를 추가하지 못했어요. 🫠

우선 제출 기한이 임박하여 PR을 먼저 생성하겠습니다.
이후에 테스트 코드를 마저 작성하여 모든 요구사항을 만족시키겠습니다!
이번 리뷰도 잘 부탁드려요~~ 🙇‍♂️ 🐿️

Junyoung-WON committed:
docs: (3단계) 기능 요구 사항 작성

Junyoung-WON committed:
feat: (3단계) 사용자 이름 입력에 대한 검증 로직 구현

- 이름은 2~5자 이내여야 한다.
- 이름에는 숫자나 기호가 포함될 수 없다: 정규식 활용

Junyoung-WON committed:
refactor: (3단계) Layer 내부 TextField modifier의 변수명 변경

Junyoung-WON committed:
feat: (3단계) SingleLineTextInput 의 입력 안내 문구와 조건 추가

Junyoung-WON committed:
feat: (3단계) 사용자 이름에 대한 입력 검증 로직 적용

- 사용자 이름 입력 검증 로직 수정
- SignUpInteractionLayer 내에서 이름 입력 시 안내 문구를 remember로 저장

Junyoung-WON committed:
feat: (3단계) 상태 호이스팅 원칙 적용

Junyoung-WON committed:
feat: (3단계) 이메일 검증 로직 구현

Junyoung-WON committed:
feat: (3단계) 이메일 검증 로직 적용

Junyoung-WON committed:
feat: (3단계) 비밀번호 검증 로직 구현

- 8~16자 이내
- 영문과 숫자 포함
- 비밀번호 확인 입력은 비밀번호와 동일해야함

Junyoung-WON committed:
feat: (3단계) 비밀번호 검증 로직 적용

Junyoung-WON committed:
feat: (3단계) 비밀번호 형식 검증 로직 수정

- 입력 길이에 대한 검증은 따로 수행하고 있으므로, 정규식에 포함된 입력 길이 조건을 제거한다.

Junyoung-WON committed:
refactor: 학습 테스트 파일을 study 패키지로 이동

Junyoung-WON committed:
refactor: (3단계) 회원가입 입력 값들의 검증 로직 테스트 코드 작성

- 사용자 이름 테스트: 입력 길이, 입력 형식 테스트
- 이메일 테스트: 입력 형식 테스트
- 비밀번호 테스트: 입력 길이, 입력 형식, 비밀번호 확인 테스트

Junyoung-WON committed:
docs: (3단계) 기능 요구 사항 만족 여부 확인

Junyoung-WON committed:
docs: (4단계) 기능 요구 사항 작성

Junyoung-WON committed:
feat: (4단계) 버튼의 enabled 속성을 외부에서 받도록 변경

Junyoung-WON committed:
feat: (4단계) 사용자 이름을 스스로 검증하는 Username 도메인 작성

Junyoung-WON committed:
feat: (4단계) Username 도메인을 사용한 검증 로직 작성 및 적용

Junyoung-WON committed:
feat: (4단계) 이메일 입력을 스스로 검증하는 도메인 Email 생성

Junyoung-WON committed:
feat: (4단계) Email 모델을 이용한 입력 검증 로직 사용

Junyoung-WON committed:
feat: (4단계) 비밀번호의 유효성을 검증하는 Password 모델 생성

Junyoung-WON committed:
feat: (4단계) Password 를 활용하여 검증 로직을 처리하도록 변경

Junyoung-WON committed:
feat: (4단계) 도메인 모델을 data class로 변경 및 도메인을 저장하도록 변경

Junyoung-WON committed:
feat: (4단계) 모든 입력값이 유효성에 따라 회원가입 버튼의 enable 설정

Junyoung-WON committed:
refactor: (4단계) 비밀번호 확인 검증 메서드 수정하여 적용

Junyoung-WON committed:
refactor: (4단계) 확장함수 활용

Junyoung-WON committed:
refactor: (4단계) 문자열 리소스 활용

Junyoung-WON committed:
refactor: (4단계) SignUpValidator를 활용하여 검증 테스트를 하도록 변경

Junyoung-WON committed:
refactor: (4단계) 사용하지 않는 검증 로직 제거

Junyoung-WON committed:
style: (4단계) ktlint 적용

Junyoung-WON committed:
docs: (4단계) 기능 요구 사항 만족 여부 확인

- 모든 로직에 대한 테스트 코드를 추가하지 못함
- 각 도메인별 유효성 검증 테스트 작성 필요

kkosang commented on README.md:
```diff
@@ -109,3 +109,22 @@ class LayoutBasicsTest {
+
+## 기능 요구 사항
+- [x] 디자인 시안을 참고하여 모든 필드가 에러 없이 채워진 경우에만 Sign up 버튼을 활성화한다.
+- [x] (선택사항) Sign up 버튼을 클릭하면 회원가입 완료 스낵바가 노출된다.
```
옹 선택 요구사항도 구현하셨군요 👍

Junyoung-WON committed:
refactor: (4단계) SignUpValidator 검증 메서드 리팩터링

- if 문 대신 when 문 활용

Junyoung-WON committed:
refactor: (4단계) supportingText 의 상태 값만을 받아오도록 변경

- 상태 호이스팅 원칙, UDF 원칙에 따라 외부에서 supportingText 값을 받도록 변경
- SingleLineTextInput 의 변경에 따라 테스트코드 수정

Junyoung-WON committed:
refactor: (4단계) 스낵바 적용 및 컴포지션 최적화 진행

스낵바 활용
- SnackBarHostState와 코루틴 스코프를 활용하여 SnackBarHost 구성
- Box 레이아웃을 활용하여 스낵바를 레이아웃 위에 출력

Recomposition Issue 해결
- 하나의 TextField 입력 값 변경에도 다른 TextField들이 재구성되는 현상 제거
- 상태 호이스팅 원칙, UDF 원칙 적용

Junyoung-WON committed:
refactor: (4단계) 토스트 대신 스낵바 적용

Junyoung-WON committed:
refactor: (4단계) 컴포즈 컨벤션 적용하여 구성요소 순서 변경

Junyoung-WON committed:
refactor: (4단계) 사용하지 않는 import 제거

Junyoung-WON committed:
style: (4단계) ktlint 적용

kkosang commented on app/src/main/java/nextstep/signup/domain/Email.kt:
```diff
@@ -0,0 +1,14 @@
+package nextstep.signup.domain
+
+data class Email(val value: String = "") {
```
전체적으로 상수화를 깔끔하게 해주셨는데 이 부분에서도 챙겨주면 좋을 것 같아요 !

Junyoung-WON replied:
`DefaultValue` 와 같은 이름으로 상수화를 하려 했는데, 빈 값이라는 것을 제대로 나타내지 못할 것 같아 상수화를 하지 않았습니다.
하지만 지금 보니 `EmptyEmail` 처럼 조금 더 명시적으로 둘 수 있겠네요!

Junyoung-WON replied:
비어있는 이메일 값이라는 것을 나타내도록 `EMPTY_EMAIL`라는 상수를 만들어 적용하였습니다!
```kotlin
data class Email(val value: String = EMPTY_EMAIL) {
    // ...
    companion object {
        private const val EMPTY_EMAIL = ""
        // ...
    }
}
```
이메일 뿐 아니라 비밀번호, 사용자 이름 도메인 모델에도 모두 적용하였어요~

피드백 반영 커밋 : [6d82936](https://github.com/woowacourse/android-signup/pull/38/commits/6d82936b617b1dea52e8f64b4c452c88712f66d0)(Email), [fbc8d10](https://github.com/woowacourse/android-signup/pull/38/commits/fbc8d103176bd64758e636d48eca8a1ceb9c10d9)(Password), [6a62fc7](https://github.com/woowacourse/android-signup/pull/38/commits/6a62fc76e37633846d9e0dc2eeedab17b0e3c821)(Username)

kkosang commented on app/src/androidTest/java/nextstep/signup/EmailInputValidationTest.kt:
전체적으로 테스트 작성 잘하셨네요 !!
입력값에 대한 유효성 검사를 모두 성공하면 버튼이 활성화 되는 시나리오에 대해서도 테스트 해보면 좋을 것 같아요~!

Junyoung-WON replied:
이번에는 꼭...! 버튼 활성화 테스트!!! 도메인 테스트도 같이!
작성해서 제출하겠습니다!!!

Junyoung-WON replied:
드디어,,, 테스트가 추가되었습니다 😭
**도메인 테스트(검증 로직 테스트)**와 꼬상이 피드백 주신 **입력 값 유효성에 따른 버튼 활성화 테스트**를 추가하였습니다!

피드백 반영 커밋 :
- 도메인 테스트
  - Email : [4ab98ad](https://github.com/woowacourse/android-signup/pull/38/commits/4ab98ad72c4fd43bf0035874870b74241a5d89da)
  - Password : [608e5e2](https://github.com/woowacourse/android-signup/pull/38/commits/608e5e29a5eb7aa6bfae355104da5e52377e0a25)
  - PasswordConfirm : [02dda9f](https://github.com/woowacourse/android-signup/pull/38/commits/02dda9f00d935825b86d75be6e5a8d879501cf07)
  - Username : [d8c8159](https://github.com/woowacourse/android-signup/pull/38/commits/d8c8159932063498b3579ec336878cac5bc71769)
- 입력 값의 유효성 검사 결과에 따라 버튼의 활성화/비활성화 테스트 : [ba9f919](https://github.com/woowacourse/android-signup/pull/38/commits/ba9f9193693416fbde05c6cca1a694a31b0fde27)

kkosang replied:
테스트를 엄청 꼼꼼하게 해주셨네요 !! 굿 👍💯

kkosang commented on app/src/main/java/nextstep/signup/ui/signup/SignUpLayouts.kt:
```diff
@@ -1,27 +1,77 @@
+fun SignUpLayout() {
+    val snackbarHostState by remember { mutableStateOf(SnackbarHostState()) }
+    val coroutineScope = rememberCoroutineScope()
+    val localContextResources = LocalContext.current.resources
```
showSnackbar에서 사용하기 위해서 localContext를 사용하셨네요 !
LocalContext와 Context의 차이가 무엇인가요?? 🤔

Junyoung-WON replied:
나중에 꼭 답변드릴게요...!! 🥹

kkosang commented on app/src/main/java/nextstep/signup/ui/signup/SignUpLayouts.kt:
```diff
@@ -32,60 +82,87 @@ fun SignUpTitle() {
+    var password by remember { mutableStateOf(Password()) }
+    var passwordConfirm by remember { mutableStateOf("") }
+    val isFormatValid by remember {
+        derivedStateOf {
```
derivedStateOf 활용 좋아요 👍

kkosang commented on app/src/main/java/nextstep/signup/domain/Password.kt:
```diff
@@ -0,0 +1,20 @@
+    val isValidFormat: Boolean
+        get() = value.validatePasswordFormat()
+
+    private fun String.validatePasswordLength(): Boolean = length in PASSWORD_LEGNTH
```
확장함수를 사용해서 표현하셨네요 !!
호두는 어떤 경우에 확장함수를 사용하시나요??

Junyoung-WON replied:
함수 내부에서 처리하는 연산이 대부분 매개변수로 받은 객체의 동작(객체의 메서드)으로 이루어질 때 자주 활용하는 것 같아요!
검증 값을 바탕으로 적절한 `supportingText`를 반환해주는 `SignUpValidator`의 메서드에서도 비슷한 동작이 많아 활용했습니다~
어떤 객체가 주체가 되어 행하는 동작인지도 잘 나타나고, 가독성 측면에서도 좋아보여서 적용할 수 있다면 최대한 적용하고 있어요!

```kotlin
// 기존 함수
@Composable
fun validateUsername(username: Username): String? {
    return when(username) {
        value.isEmpty() -> null
        isValidFormat.not() -> stringResource(id = R.string.signup_support_username_invalid_format)
        isValidLength.not() -> stringResource(id = R.string.signup_support_username_invalid_length)
        else -> null
    }
}

// 확장 함수
@Composable
fun Username.validateUsername(): String? {
    return when {
        value.isEmpty() -> null
        isValidFormat.not() -> stringResource(id = R.string.signup_support_username_invalid_format)
        isValidLength.not() -> stringResource(id = R.string.signup_support_username_invalid_length)
        else -> null
    }
}
```

Junyoung-WON replied:
그리고 "확장" 함수라는 의미에 알맞게, 객체가 행해야하는 주요 역할이 아니면서도, **특정 부분에서 활용되는 비즈니스 로직**인 경우(지금 구현의 경우에는 검증 결과에 따라 에러 메시지를 반환하는 것이 될 것 같네요!)에 **기능을 확장한다라는 의미**에서 활용하는 것 같습니다!

kkosang replied:
옹 저는 확장함수를 잘 사용하지 않는 편인데

> 함수 내부에서 처리하는 연산이 대부분 매개변수로 받은 객체의 동작(객체의 메서드)으로 이루어질 때 자주 활용하는 것 같아요!

호두는 명확하게 사용하는 이유가 있었군요 ! `매개변수로 받은 객체의 동작`이라는 이유가 저도 사용하고 싶게 만드는 이유네요  😁 👍

kkosang commented on app/src/main/java/nextstep/signup/ui/common/textfield/SingleLineTextInput.kt:
```diff
@@ -47,5 +33,18 @@ fun SingleLineTextInput(
                 focusedIndicatorColor = MaterialTheme.colorScheme.secondary,
                 focusedLabelColor = MaterialTheme.colorScheme.secondary,
             ),
+        isError = supportingText.isNullOrBlank().not(),
```
`isNullOrBlank()` 라는것도 있군요 !! 👍

Junyoung-WON replied:
맞아요!
`isNullOrEmpty()` 와 동작은 비슷하지만, **공백 문자로만 이루어져있는지도 함께 판단**해주어서 종종 유용하게 활용하는 확장함수입니다!

```kotlin
@InlineOnly
public inline fun CharSequence?.isNullOrBlank(): Boolean
```
> Returns true if this nullable char sequence is either null or empty or consists solely of whitespace characters.

kkosang commented on app/src/main/java/nextstep/signup/ui/signup/SignUpValidator.kt:
```diff
@@ -0,0 +1,48 @@
+
+object SignUpValidator {
+    @Composable
+    fun Username.validateUsername(): String? {
```
저는 뭔가 함수명이 조금 어색하게 느껴집니다.
함수 내부의 구현을 보면 Username의 유효성에 따라서 오류메시지를 반환해주고 있는데  `validateUsername`만 봤을 때 오류 메시지를 반환해준다는 느낌보다 유효한 이름을 반환해준다는 느낌을 받았어요 !

Junyoung-WON replied:
그러네요...! 꼬상의 말대로, 유효성에 따라 사용자에게 보여줄 적절한 메시지를 반환해주는 일을 수행하는데
메서드명만 놓고 본다면 유효한 이름을 반환해주거나, "검증한다"라는 동작을 수행하는 느낌이 강하네요!

다른 메서드 명으로 바꾸어서 더 명시적으로 나타내볼게요!

Junyoung-WON replied:
`validate~` 라는 메서드 명에서, `getValidationMessage` 으로 검증 메시지를 반환한다는 의미를 나타내도록 변경하였습니다!

피드백 반영 커밋 : [e2cf417](https://github.com/woowacourse/android-signup/pull/38/commits/e2cf417591e7c8f7fb6e7d015eaf27b66dfceec0)

kkosang reviewed (changes requested):
안녕하세요 호두 !

미션 3,4 단계 구현하시느라 고생 많으셨습니다 ! 💯
미션을 진행하면서 리컴포지션을 고려하여 설계를 하신 것 같군요 👍
몇 가지 코멘트를 남겨놓았으니 확인 부탁드려요~

Junyoung-WON committed:
refactor: (4단계) Email 내부 원시값들의 상수화 진행

- 프로퍼티 디폴트 값 상수화 진행

Junyoung-WON committed:
refactor: (4단계) Password 내부 원시값들의 상수화 진행

- 프로퍼티 디폴트 값, 글자 수 상수화 진행
- 상수값 오타 수정

Junyoung-WON committed:
refactor: (4단계) Username 내부 원시값들의 상수화 진행

- 프로퍼티 디폴트 값, 글자 수 상수화 진행
- 상수값 오타 수정

Junyoung-WON committed:
feat: (4단계) 비밀번호 확인 값을 검증하는 도메인 모델 추가

Junyoung-WON committed:
refactor: (4단계) 검증 결과 메시지를 반환하는 메서드명 변경

- 검증 메시지를 반환하는 동작을 명시적으로 나타냄
- PasswordConfirm 활용

Junyoung-WON committed:
refactor: (4단계) 회원가입 화면에서 사용하는 컴포저블을 정의

- 재사용 가능한 컴포저블만을 사용하기보다, 현재 화면에서의 사용되는 목적에 맞게 컴포저블을 명시적으로 나타내어, 가독성을 높이고 명확한 역할을 부여

Junyoung-WON committed:
refactor: (4단계) 회원가입 화면에서 사용하는 컴포저블 적용

Junyoung-WON committed:
test: (4단계) Email 검증 도메인 테스트 작성

- 로컬 파트 : 대소문자, 숫자, '.', '_', '%', '+', '-', 1글자 이상
- 도메인 이름 : 대소문자, 숫자, '.', '-', 1글자 이상
- 최상위 도메인 : 대소문자, 2~6글자

Junyoung-WON committed:
test: (4단계) Password 검증 도메인 테스트 작성

- 글자 수 : 8~16자
- 형식 : 알파벳과 숫자가 최소 하나 이상 포함되어야 한다

Junyoung-WON committed:
test: (4단계) Username 검증 도메인 테스트 작성

- 글자 수 : 2~5자
- 형식 : 대소문자와 한글 이외의 문자는 유효하지 않다

Junyoung-WON committed:
test: (4단계) PasswordConfirm 검증 도메인 테스트 작성

- 형식 : 입력한 비밀번호와 동일하면 유효하다

Junyoung-WON committed:
test: (4단계) 사용자 입력에 따른 유효성 및 회원가입 버튼 활성화 여부 테스트 작성

Junyoung-WON committed:
style: (4단계) ktlint 적용

Junyoung-WON committed:
docs: (4단계) 요구 사항 만족 여부 확인

Junyoung-WON commented:
안녕하세요, 꼬상!
많이 배려해주신 덕분에, 미션에 조금 더 여유있게 투자할 수 있었어요!
정말 감사합니다...🙇‍♂️

꼬상이 피드백주신 리팩터링, 테스트와 함께, UI 로직을 조금 더 리팩터링해보고싶어 **컴포저블을 분리**해보았습니다!
예를 들어 Column 안에서 단순히 SingleLineTextInput 으로 두는 것이 아니라,
각각의 TextInput을 컴포저블로 감싸서 어떤 입력에 대한 TextInput인지 명시적으로 나타내었습니다.

아래에 주요 변경사항 별 커밋 내역을 정리했습니다!

### 주요 변경사항 별 커밋 내역
- 원시값 상수화 진행
  - Email : [6d82936](https://github.com/woowacourse/android-signup/pull/38/commits/6d82936b617b1dea52e8f64b4c452c88712f66d0)
  - Password : [fbc8d10](https://github.com/woowacourse/android-signup/pull/38/commits/fbc8d103176bd64758e636d48eca8a1ceb9c10d9)
  - Username : [6a62fc7](https://github.com/woowacourse/android-signup/pull/38/commits/6a62fc76e37633846d9e0dc2eeedab17b0e3c821)
- 비밀번호 확인 도메인 모델 추가 : [b654338](https://github.com/woowacourse/android-signup/pull/38/commits/b654338eff6e4aeb1cbc7f838407f61d54bd4eac)
- 검증 결과 메시지를 반환하는 메서드 명 변경 : [e2cf417](https://github.com/woowacourse/android-signup/pull/38/commits/e2cf417591e7c8f7fb6e7d015eaf27b66dfceec0)
- 회원가입 화면의 컴포저블 분리 및 적용 : [f6bd1a9](https://github.com/woowacourse/android-signup/pull/38/commits/f6bd1a98f253ed377d5ade4536eb662721221adb), [7472ccf](https://github.com/woowacourse/android-signup/pull/38/commits/7472ccf97921a73866a83fdf6f5f66470ca6d2d0)
- 도메인(검증 로직) 테스트 작성
  - Email : [4ab98ad](https://github.com/woowacourse/android-signup/pull/38/commits/4ab98ad72c4fd43bf0035874870b74241a5d89da)
  - Password : [608e5e2](https://github.com/woowacourse/android-signup/pull/38/commits/608e5e29a5eb7aa6bfae355104da5e52377e0a25)
  - PasswordConfirm : [02dda9f](https://github.com/woowacourse/android-signup/pull/38/commits/02dda9f00d935825b86d75be6e5a8d879501cf07)
  - Username : [d8c8159](https://github.com/woowacourse/android-signup/pull/38/commits/d8c8159932063498b3579ec336878cac5bc71769)
- 입력 값의 유효성 검사 결과에 따라 버튼의 활성화/비활성화 테스트 : [ba9f919](https://github.com/woowacourse/android-signup/pull/38/commits/ba9f9193693416fbde05c6cca1a694a31b0fde27)

kkosang commented on app/src/test/java/nextstep/signup/domain/EmailTest.kt:
```diff
@@ -0,0 +1,133 @@
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+
+class EmailTest {
```
헉 도메인 테스트를 엄청 여러가지 시나리오에 대해서 테스트 해주셨네요 💯

kkosang commented on app/src/androidTest/java/nextstep/signup/SignUpInteractionLayerTest.kt:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    @Test
+    fun `모든_입력_값이_비어있는_초기에는_회원가입_버튼이_비활성화되어_있다`() {
```
버튼 활성화 테스트 좋아요 💯

kkosang commented on app/src/androidTest/java/nextstep/signup/SignUpInteractionLayerTest.kt:
```diff
@@ -0,0 +1,208 @@
+        composeTestRule.setContent {
+            val modifier = Modifier.fillMaxWidth()
+            SignUpInputs(
+                modifier = modifier.testTag(SIGN_UP_INPUTS_TAG),
```
modifier에 testTag를 만들어서 달아둘 수도 있군요... 
저도 테스트할 때 유용하게 써야겠어요 ~~

kkosang commented:
비밀번호 확인하는 검증 테스트 2개가 실패하고 있어요..! 
비밀번호가 일치해도 supportingText가 보이고 있답니다.. 😱
<img width="265" alt="스크린샷 2024-10-23 오후 8 14 49" src="https://github.com/user-attachments/assets/48ba455d-f6b8-4633-a4a3-70a50f83759d">

kkosang commented on app/src/main/java/nextstep/signup/domain/Email.kt:
```diff
@@ -0,0 +1,15 @@
+    val isValidFormat: Boolean
+        get() = value.validateEmailFormat()
+
+    private fun String.validateEmailFormat(): Boolean = matches(Regex(EMAIL_REGEX))
```
Regex의 객체의 생성비용이 비싸서 matches 할 때마다 객체를 생성하면 성능이 떨어질 수 있다고 하더라구요 !
[Regex](https://velog.io/@byeongju/%EC%A0%95%EA%B7%9C%ED%91%9C%ED%98%84%EC%8B%9D%EC%9D%84-%EA%B2%80%EC%82%AC%EB%A5%BC-%EB%8D%94-%ED%9A%A8%EC%9C%A8%EC%A0%81%EC%9C%BC%EB%A1%9C-%ED%95%B4%EB%B3%B4%EC%9E%90)

kkosang reviewed (approved):
안녕하세요 호두 !!

리뷰 반영을 너무 잘 해주셨네요~ 👍
테스트 코드를 정말 꼼꼼히 작성하셨네요 !! 
회원가입 미션 하느라 고생 많았습니다. 마지막까지 화이팅입니다~
스타카토 화이팅 💪💪💪💪
